source/dem/force_chains_visualization.cc

```diff
@@ -0,0 +1,511 @@
+  // auto &ghost_adjacent_particles =
+  // container_manager.ghost_adjacent_particles; auto
+  // &local_periodic_adjacent_particles =
+  //  container_manager.local_periodic_adjacent_particles;
+  // auto &ghost_periodic_adjacent_particles =
+  //  container_manager.ghost_periodic_adjacent_particles;
+  // auto &ghost_local_periodic_adjacent_particles =
+  //  container_manager.ghost_local_periodic_adjacent_particles;
+
+  // Define local variables which will be used within the contact calculation
+  //  Namely: normal and tangential contact forces, tangential and rolling
+  //  torques, normal unit vector of the contact and contact relative velocity
+  //  in the normal direction
+  Tensor<1, 3> normal_unit_vector;
+  Tensor<1, 3> normal_force;
+  Tensor<1, 3> tangential_force;
+  Tensor<1, 3> particle_one_tangential_torque;
+  Tensor<1, 3> particle_two_tangential_torque;
+  Tensor<1, 3> rolling_resistance_torque;
+  double       normal_relative_velocity_value;
+  Tensor<1, 3> tangential_relative_velocity;
+
+  // Contact forces calculations of local-local and local-ghost particle
+  // pairs are performed in separate loops
+
+  // Looping over local_adjacent_particles values with iterator
+  // adjacent_particles_list
+  for (auto &&adjacent_particles_list :
+       local_adjacent_particles | boost::adaptors::map_values)
+    {
+      if (!adjacent_particles_list.empty())
+        {
+          // Gather information about particle 1 and set it up.
+          auto first_contact_info = adjacent_particles_list.begin();
+          auto particle_one       = first_contact_info->second.particle_one;
+          auto particle_one_properties = particle_one->get_properties();
+
+
+          // Fix particle one location for 2d and 3d
+          Point<3> particle_one_location = [&] {
+            if constexpr (dim == 3)
+              {
+                return particle_one->get_location();
+              }
+            if constexpr (dim == 2)
+              {
+                return (point_nd_to_3d(particle_one->get_location()));
+              }
+          }();
+
+          for (auto &&contact_info :
+               adjacent_particles_list | boost::adaptors::map_values)
+            {
+              // Getting information (location and properties) of particle 2 in
+              // contact with particle 1
+              auto particle_two            = contact_info.particle_two;
+              auto particle_two_properties = particle_two->get_properties();
+
+              // Get particle 2 location in dimension independent way
+              Point<3> particle_two_location = [&] {
+                if constexpr (dim == 3)
+                  {
+                    return particle_two->get_location();
+                  }
+                if constexpr (dim == 2)
+                  {
+                    return (point_nd_to_3d(particle_two->get_location()));
+                  }
+              }();
```

Comment:
Same here

Reply:
Please don't leave it like it is now, this is the right way.